contributors/design-proposals/kubelet-client-certificate-bootstrap-rotation.md

```diff
@@ -0,0 +1,90 @@
+Certificate Signing Request API to sign a new key.
+
+## Use Cases
+
+1. Kubelet starts for the first time on a new node.
+2. Kubelet certificate approaches expiration.
+
+## Proposed Design
+
+### During Kubelet Boot Sequence
+
+1. Look on disk for an existing cert/key pair managed by the certificate
+   manager.
+1. If there is an existing cert/key pair, load them.
+1. If there is no existing cert/key pair, look for cert/key data specified in
+   the kubelet config file:
+    - values encoded in the kubeconfig (CertData, KeyData, CAData)
+    - file references in the kubeconfig (CertFile, KeyFile, CAFile)
+1. If the certificate is a bootstrap certificate, use the certificate to:
+    - generate a key
+    - create a certificate signing request
+    - request a signed certificate from the API server
+    - wait for a response
+    - Store the new cert/key pairs in the directory specified by `--cert-dir`,
+      or the default value if it is unspecified.
+
+### As Expiration Approaches
+
+Rotating a Kubelet client certificate will work by generating a new private key,
+issuing a new Certificate Signing Request to the API Server, safely updating the
+cert/key pair on disk, begin using the new cert/key pair.
+
+1. Store the new cert/key pairs in the directory specified by `--cert-dir`, or the
+   default value if it is unspecified. This will allow the kubelet to have a
+   place for storing the multiple cert/key pairs that it might have available at
+   any given moment (because of a rotation in progress).
+    - When cert/key files are specified in the kubeconfig, these will be used if
+      a newer, rotated, cert/key pair does not exist.
+1. There will be a feature gate to enable certificate bootstrapping and
+   rotation, `RotateKubeletClientCertificate`
+1. Centralize certificate access within the kubelet code to the
+   CertificateManager. Whenever rotation is enabled, CertificateManager will be
+   responsible for:
+    - Providing the correct certificate to use for establishing TLS connections.
+    - Generating new private keys and requesting new certificates when the
+      current certificate approaches expiry.
+    - Since certificates can rotate at any time, all other parts of the kubelet
+      should ask the CertificateManager for the correct certificate each time a
+      certificate is used. No certificate caching except by the
+      CertificateManager.
+        - TLS connections should prefer to set the
+          [`GetCertificate`](https://golang.org/pkg/crypto/tls/#Config) and
+          [`GetClientCertificate`](https://golang.org/pkg/crypto/tls/#Config)
+          callbacks so that the connection dynamically requests the certificate
+          as needed.
+    - Recovering from kubelet crashes or restarts that occur while certificate
+      transitions are in flight (request issued, but not yet signed, etc)
+1. The RBAC role for nodes will be updated to allow nodes to request new
```

Comment:
the role already allows requesting - https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L709-L717